pubsub.go

```diff
@@ -259,28 +267,53 @@ func (p *PubSub) processLoop(ctx context.Context) {
 		p.peers = nil
 		p.topics = nil
 	}()
+
 	for {
 		select {
-		case s := <-p.newPeers:
-			pid := s.Conn().RemotePeer()
-			ch, ok := p.peers[pid]
+		case pid := <-p.newPeers:
+			_, ok := p.peers[pid]
 			if ok {
-				log.Error("already have connection to peer: ", pid)
-				close(ch)
+				log.Warning("already have connection to peer: ", pid)
+				continue
 			}
 
 			messages := make(chan *RPC, 32)
-			go p.handleSendingMessages(ctx, s, messages)
 			messages <- p.getHelloPacket()
-
+			go p.handleNewPeer(ctx, pid, messages)
 			p.peers[pid] = messages
 
+		case s := <-p.newPeerStream:
+			pid := s.Conn().RemotePeer()
+
+			_, ok := p.peers[pid]
+			if !ok {
+				log.Warning("new stream for unknown peer: ", pid)
+				s.Reset()
+				continue
+			}
+
 			p.rt.AddPeer(pid, s.Protocol())
 
+		case pid := <-p.newPeerError:
+			delete(p.peers, pid)
```

Comment:
There's a race here where:

1. We have an old zombie connection that doesn't know it's dead yet.
2. We get a new connection.

With the current logic, we may still try to open a stream using the old connection. We'll then never try to create one with the new connection.

Unfortunately, this is a pretty common situation when a connection dies and we replace it immediately.

Maybe we should drop this channel and send a notification along `peerDead` instead? If we still think we're connected, we'll continue to retry until we realize we're actually disconnected.

Reply:
This is only used when the stream failed to open; we can't pass it through the `peerDead` channel because the peer may simply not support the pubsub protocols.

Reply:
The problem here is if the stream fails to open because the connection was closed (if it was a duplicate connection). 
Is there a way to discriminate whether the failure was an unsupported protocol?
In this case we can notify on `peerError` for unsupported protocol and `peerDead` for connection closed errors.

Reply:
named error in `go-multistream.ErrNotSupported` can be used for this per discussion on irc.

Reply:
done.